# *NanoPush* 

### Development of an open-source, AI-driven DNA methylation caller on CpG dinucleotides for single-step epigenetic analysis of _de-novo_ sequenced organisms by ’3rd generation’ Nanopore sequencing technology

## PhD Rotation Project Abstract
The post-replicative DNA modifications are highly conserved across species and play a key role in development and ability of organisms to adapt to their environment through epigenetic mechanisms. The development of Oxford Nanopore Technologies (ONT) MinION ‘3rd generation’ sequencing technology allows for direct characterisation of the modified bases with single-base resolution. Nanopore uses a methodology in which electrical current is disrupted as DNA or RNA molecule is pulled through a nanopore, and DNA or RNA sequence is determined based on the unique signatures of unmodified and modified bases. In this project, we developed a novel single-step DNA methylation caller as a tool to detect cytosine modifications in CpG dinucleotide context using machine learning approaches (support vector machine). We aimed to develop our base calling algorithm to provide an improvement to current ONT methodologies as those currently do not allow for detection of single organism differences in methylation profiles of _de-novo_ sequenced social organisms, which respond robustly to environmental factors, including nutritional control.

### Project Status

This project has been terminated upon completion of the PhD rotation in the laboratory. The development of the code has stopped completely. If you'd like to get in touch, please contact me directly via [email](mailto:kristina.smith.ulicna@gmail.com) or check out my [LinkedIn](https://www.linkedin.com/in/kristinaulicna/) profile for a more detailed description of the project. 
